core/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -79,54 +66,41 @@
 
 import static java.util.Collections.emptyMap;
 import static java.util.Collections.unmodifiableMap;
-import static org.elasticsearch.cluster.SnapshotsInProgress.completed;
 
 /**
  * This service runs on data and master nodes and controls currently snapshotted shards on these nodes. It is responsible for
  * starting and stopping shard level snapshots
  */
-public class SnapshotShardsService extends AbstractLifecycleComponent implements ClusterStateApplier, IndexEventListener {
-
-    public static final String UPDATE_SNAPSHOT_ACTION_NAME = "internal:cluster/snapshot/update_snapshot";
+public class SnapshotShardsService extends AbstractLifecycleComponent implements ClusterStateListener, IndexEventListener {
 
     private final ClusterService clusterService;
 
     private final IndicesService indicesService;
 
     private final SnapshotsService snapshotsService;
 
-    private final TransportService transportService;
-
     private final ThreadPool threadPool;
 
     private final Lock shutdownLock = new ReentrantLock();
 
     private final Condition shutdownCondition = shutdownLock.newCondition();
 
     private volatile Map<Snapshot, SnapshotShards> shardSnapshots = emptyMap();
-
-    private final SnapshotStateExecutor snapshotStateExecutor = new SnapshotStateExecutor();
+    private final Client client;
 
     @Inject
-    public SnapshotShardsService(Settings settings, ClusterService clusterService, SnapshotsService snapshotsService, ThreadPool threadPool,
-                                 TransportService transportService, IndicesService indicesService) {
+    public SnapshotShardsService(Settings settings, ClusterService clusterService, SnapshotsService snapshotsService,
+                                 ThreadPool threadPool, IndicesService indicesService, Client client) {
         super(settings);
         this.indicesService = indicesService;
         this.snapshotsService = snapshotsService;
-        this.transportService = transportService;
         this.clusterService = clusterService;
         this.threadPool = threadPool;
+        this.client = client;
         if (DiscoveryNode.isDataNode(settings)) {
-            // this is only useful on the nodes that can hold data
-            // addLowPriorityApplier to make sure that Repository will be created before snapshot
-            clusterService.addLowPriorityApplier(this);
+            // this is only useful on the nodes that can hold data.
+            clusterService.addListener(this);
```

Comment:
Could you add a comment here why this is needs to be a listener and not an applier and why it's ok to have this as `addListener` and not `addLast`?

Reply:
The test `DedicatedClusterSnapshotRestoreIT#testSnapshotWithStuckNode` was failed with an `Applier` but passed with a `Listener`. However, I don't really know the root cause.

Reply:
@imotov, I think I have figured out the root cause. A Listener is called after the new state is set, while an Applier is called before that. In `SnapshotShardsService`, we call `#syncShardStatsOnNewMaster` -> `#updateIndexShardSnapshotStatus` which in turn uses a TransportMasterNodeAction. The TransportMasterNodeAction accesses the state of the cluster directly which may not be available yet if it's invoked from the Applier.


```
Caused by: java.lang.AssertionError: should not be called by a cluster state applier. reason [the applied cluster state is not yet available]
```

This explains why `#testSnapshotWithStuckNode` was failed with an Applier but passed with a Listener.